x-pack/plugins/enterprise_search/public/applications/app_search/components/source_engines/source_engines_logic.ts

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { kea, MakeLogicType } from 'kea';
+
+import { Meta } from '../../../../../common/types';
+
+import { flashAPIErrors } from '../../../shared/flash_messages';
+import { HttpLogic } from '../../../shared/http';
+import { EngineLogic } from '../engine';
+import { EngineDetails } from '../engine/types';
+
+interface EnginesAPIResponse {
+  results: EngineDetails[];
+  meta: Meta;
+}
+
+interface SourceEnginesLogicValues {
+  dataLoading: boolean;
+  sourceEngines: EngineDetails[];
+}
+
+interface SourceEngineLogicActions {
+  fetchSourceEngines: () => void;
+  setSourceEngines: (
+    sourceEngines: SourceEnginesLogicValues['sourceEngines']
+  ) => { sourceEngines: SourceEnginesLogicValues['sourceEngines'] };
+}
+
+export const SourceEnginesLogic = kea<
+  MakeLogicType<SourceEnginesLogicValues, SourceEngineLogicActions>
+>({
+  path: ['enterprise_search', 'app_search', 'curations_logic'],
+  actions: () => ({
+    fetchSourceEngines: true,
+    setSourceEngines: (sourceEngines) => ({ sourceEngines }),
+  }),
+  reducers: () => ({
+    dataLoading: [
+      true,
+      {
+        setSourceEngines: () => false,
+      },
+    ],
+    sourceEngines: [
+      [],
+      {
+        setSourceEngines: (_, { sourceEngines }) => sourceEngines,
+      },
+    ],
+  }),
+  listeners: ({ actions }) => ({
+    fetchSourceEngines: () => {
+      const { http } = HttpLogic.values;
+      const { engineName } = EngineLogic.values;
+
+      let enginesAccumulator: EngineDetails[] = [];
+
+      // We need to recursively fetch all source engines because we put the data
+      // into an EuiInMemoryTable to enable searching
+      const recursiveFetchSourceEngines = async (page = 1) => {
+        try {
+          const { meta, results }: EnginesAPIResponse = await http.get(
+            `/api/app_search/engines/${engineName}/source_engines`,
+            {
+              query: {
+                'page[current]': page,
+                'page[size]': 25,
+              },
+            }
+          );
+
+          enginesAccumulator = [...enginesAccumulator, ...results];
+
+          if (page >= meta.page.total_pages) {
+            actions.setSourceEngines(enginesAccumulator);
```

Comment:
Just curious, not a change request - any particular reason this is `addSourceEngine` in `MetaEnginesTableLogic` but `setSourceEngines` here? I don't have a strong preference, but if I had to choose I like the simplicity of `setSourceEngines` 👍

Reply:
Ah okay so the difference is that in `MetaEnginesTableLogic` the value `sourceEngines` is a `{ [id: string]: EngineDetails[] };` but in `SouceEnginesLogic` the value is just `EngineDetails[];`.  MetaEnginesTableLogic tracks multiple different engines (since the MetaEnginesTable has a paginated list of all meta engines).  Here, we only care about a single engine.  So In `MetaEnginesTableLogic`, we call `actions.addSourceEngines({ [engineName]: enginesAccumulator });` to add the info for a single engine to the collection of all engines, but in `SourceEnginesLogic` we call `  actions.setSourceEngines(enginesAccumulator);` since it's just this engine.

Maybe one of these should be renamed.   Very open to suggestions!

Reply:
Ohh interesting, I hadn't fully grasped that distinction 🤔  We can definitely evaluate if/when we do a shared `fetchSourceEngines` or base logic file!

Thinking aloud - I wonder if we could key a base `SourceEnginesLogic` based on the meta engine ID, so that a single MetaEnginesTableLogic isn't responsible for every single row of source engines, and that way we can share the logic between the two views...